### DISCUSSION TITLE
正文参考文献的格式问题

### DISCUSSION THREAD

**Discussion by obstacle213** [sjtug/SJTUThesis#789] [提问]
使用的是Tex live，Windows版本

正文中引用参考文献使用\cite{}语句，出现的格式如图中红框所示 
（图为模板math_and_citations中的例子）
<img width="277" alt="1650185158(1)" src="https://user-images.githubusercontent.com/55470875/163707381-1171a2a5-eb79-48db-8103-8b88c7d40354.png">

如何实现引用格式为上标[1],[2]这样的格式呢？
我未在main和setup格式中进行过修改，求大佬解答疑惑，谢谢~

**Comment by skyzh** [comment]
用的是哪个版本？编译方式是？

**Reply by obstacle213** [reply]
Texstudio本地编译的

**Reply by LogCreative** [reply]
按照README设置一下编辑器 更改编译方式为latexmk 这种情况一般由没有进行正常的多次编译导致

**Comment by obstacle213** [comment]
我按照README如下设置的，应该是没有什么问题的
<img width="424" alt="1650189458(1)" src="https://user-images.githubusercontent.com/55470875/163709604-7e426f8c-bbbc-4a4d-a976-594f92332a2f.png">
<img width="426" alt="1650189580(1)" src="https://user-images.githubusercontent.com/55470875/163709671-d41cbce3-e0b3-4b06-9cb0-8f29884f036a.png">

**Reply by LogCreative** [reply]
那你试着将默认文献工具改为biber

**Reply by LogCreative** [reply]
不行的话 尝试使用Compile.bat看能否正常编译

**Reply by obstacle213** [reply]
我重新下载一次最新的模板也是一样。
运行Compile.bat报错
<img width="433" alt="1650247222(1)" src="https://user-images.githubusercontent.com/55470875/163742888-95b0cc33-962d-47c0-bf54-d85d08a3f57f.png">
在log文件中找到error相关的
<img width="515" alt="1650247366(1)" src="https://user-images.githubusercontent.com/55470875/163743057-8cd15cff-728c-4cb0-a076-69a896bb321a.png">
<img width="326" alt="1650247390(1)" src="https://user-images.githubusercontent.com/55470875/163743094-1ea1723b-d07c-4b36-9e02-3b2628cf06d9.png">
<img width="316" alt="1650247414(1)" src="https://user-images.githubusercontent.com/55470875/163743118-94b3953c-cb22-4e24-b5c8-3d863b44f478.png">
一些warning
<img width="361" alt="1650249911(1)" src="https://user-images.githubusercontent.com/55470875/163746253-484f6308-596e-4ef4-80c0-5acbde89e158.png">
以及很多citation undefined的warning
大佬能否帮忙看看（附件放了log）
[main.log](https://github.com/sjtug/SJTUThesis/files/8503134/main.log)

**Reply by skyzh** [reply]
顺便提供一下 main.blg？

**Reply by obstacle213** [reply]
这里好像不支持blg文件的上传，我传到网盘上
链接：https://pan.baidu.com/s/1QdQfIyHtqvyz4G6f4U_9Uw 
提取码：oef0

**Reply by skyzh** [reply]
blg 是文本文件，直接把里面的内容拷贝到这里就行了。

**Reply by obstacle213** [reply]
好的，以下是内容

```
This is BibTeX, Version 0.99d (TeX Live 2021/W32TeX)
Capacity: max_strings=200000, hash_size=200000, hash_prime=170003
The top-level auxiliary file: main.aux
I found no \citation commands---while reading file main.aux
I found no \bibdata command---while reading file main.aux
I found no \bibstyle command---while reading file main.aux
You've used 0 entries,
            0 wiz_defined-function locations,
            83 strings with 482 characters,
and the built_in function-call counts, 0 in all, are:
= -- 0
> -- 0
< -- 0
+ -- 0
- -- 0
* -- 0
:= -- 0
add.period$ -- 0
call.type$ -- 0
change.case$ -- 0
chr.to.int$ -- 0
cite$ -- 0
duplicate$ -- 0
empty$ -- 0
format.name$ -- 0
if$ -- 0
int.to.chr$ -- 0
int.to.str$ -- 0
missing$ -- 0
newline$ -- 0
num.names$ -- 0
pop$ -- 0
preamble$ -- 0
purify$ -- 0
quote$ -- 0
skip$ -- 0
stack$ -- 0
substring$ -- 0
swap$ -- 0
text.length$ -- 0
text.prefix$ -- 0
top$ -- 0
type$ -- 0
warning$ -- 0
while$ -- 0
width$ -- 0
write$ -- 0
(There were 3 error messages)
```

**Reply by skyzh** [reply]
可以先考虑 make clean 或者 Compile.bat clean 清理一下旧文件，感觉是缓存问题

**Reply by skyzh** [reply]
另外正常来讲应该用 biber 来生成参考文献数据？以下是正常的 log，感觉需要重新设置一下 TeXStudio。

```
[0] Config.pm:307> INFO - This is Biber 2.17
[0] Config.pm:310> INFO - Logfile is 'main.blg'
[53] biber-darwin:340> INFO - === Sun Apr 10, 2022, 14:56:44
[66] Biber.pm:418> INFO - Reading 'main.bcf'
[165] Biber.pm:972> INFO - Found 22 citekeys in bib section 0
[178] Biber.pm:4383> INFO - Processing section 0
[187] Biber.pm:4574> INFO - Looking for bibtex file 'bibdata/thesis.bib' for section 0
[188] bibtex.pm:1713> INFO - LaTeX decoding ...
[204] bibtex.pm:1518> INFO - Found BibTeX data source 'bibdata/thesis.bib'
[391] UCollate.pm:68> INFO - Overriding locale 'en-US' defaults 'variable = shifted' with 'variable = non-ignorable'
[391] UCollate.pm:68> INFO - Overriding locale 'en-US' defaults 'normalization = NFD' with 'normalization = prenormalized'
[391] Biber.pm:4203> INFO - Sorting list 'none/global//global/global' of type 'entry' with template 'none' and locale 'en-US'
[391] Biber.pm:4209> INFO - No sort tailoring available for locale 'en-US'
[405] bbl.pm:654> INFO - Writing 'main.bbl' with encoding 'UTF-8'
[417] bbl.pm:757> INFO - Output to main.bbl
[417] Biber.pm:130> WARN - legacy year field '1957/1990' in entry 'LSC1957' is not an integer - this will probably not sort properly.
[417] Biber.pm:132> INFO - WARNINGS: 1
```

**Reply by obstacle213** [reply]
嗯，我感觉也是biber的问题，但我选择默认文献工具为Biber时无法生成bbl文件，是不是biber哪里出了问题？

**Reply by skyzh** [reply]
先尝试清理一下临时文件，然后再用 biber 生成？

**Reply by obstacle213** [reply]
我将之前的文件夹都删除了。
重新下载了一个文件。

以下操作和结果
将默认文献工具改为了Biber。其余与README要求设置一致。
在Texstudio运行了main.tex，报错，并生成了以下文件。
<img width="189" alt="1ca3e028b6365d756a4860028915657" src="https://user-images.githubusercontent.com/55470875/163770178-f1ef2511-4024-41b4-9e5a-faec461c8700.png">

<img width="377" alt="70569d72e2f9794601c9bb7499b34b5" src="https://user-images.githubusercontent.com/55470875/163770185-1dab5528-ac7d-4796-936c-d55ce173e7fb.png">

然后将默认编译器改为了Biber，并无新文件生成并报错。
（我之前是用的BibTeX，能生成bbl文件，但是是空的）
<img width="468" alt="ab110503f86255a3687fd3a06499eab" src="https://user-images.githubusercontent.com/55470875/163770198-e547dc7f-5a83-4c62-99dd-56712bfba3f7.png">

操作上有问题么？

**Reply by LogCreative** [reply]
感觉可能 biber 的本地缓存被破坏了，在 PowerShell 中输入
```cmd
rm -r -Force $(biber --cache)
```
以清除缓存，否则就要考虑重装 TeX Live 了。可以参考 #774

**Reply by obstacle213** [reply]
我试试重装吧 好像就是软件的问题

**Reply by PhotometricStereo** [reply]
补充一个可能的原因，在修改bib的时候把其中一条的格式写错了，导致了编译bbl不通过，造成了同样的结果。T^T

**Reply by PhotometricStereo** [reply]
所以，遇到这样问题，也可能要先详细检查修改bib的item是不是正确。

**Reply by PhotometricStereo** [reply]
Win的缓存文件夹可能的位置：
C:\Users\user\AppData\Local\Temp\par-xxxxxx\cache-xxxxxxxxxxxxxxxxxxxxxx

**Reply by obstacle213** [reply]
嗯，我第二次出现这个问题是因为网上直接下载的参考文献里面有百分号，这个导致生成的文件有问题，要注意参考文献的格式和无关的符号。
